### PR TITLE
WIP: add ability to export/collect scrape_jobs

### PR DESCRIPTION
**Pull request by costela**

### CODE REVIEW

**Comment by bastelfreak** [comment]
Fixes https://github.com/voxpupuli/puppet-prometheus/issues/126

File: manifests/config.pp [hunk 1]
```diff
@@ -3,10 +3,11 @@
 class prometheus::config (
   $global_config,
   $rule_files,
-  $scrape_configs,
+  Array[Hash] $scrape_configs,
```
Comment (bastelfreak) [inline manifests/config.pp:6]: Datatypes \o/

File: manifests/daemon.pp [hunk 2]
```diff
@@ -60,22 +60,25 @@
...
+  $manage_user                    = true,
+  $extra_groups                   = [],
+  $manage_group                   = true,
+  $purge                          = true,
+  $options                        = '',
+  $init_style                     = $::prometheus::params::init_style,
+  $service_ensure                 = 'running',
+  $service_enable                 = true,
+  $manage_service                 = true,
+  Boolean $export_scrape_job      = false,
```
Comment (bastelfreak) [inline manifests/daemon.pp]: Datatypes \o/

File: manifests/init.pp [hunk 3]
```diff
@@ -113,50 +113,59 @@
...
+  $init_style                      = $::prometheus::params::init_style,
+  Hash $global_config              = $::prometheus::params::global_config,
+  Array $rule_files                = $::prometheus::params::rule_files,
+  Array $scrape_configs            = $::prometheus::params::scrape_configs,
+  Array $remote_read_configs       = $::prometheus::params::remote_read_configs,
+  $alerts                          = $::prometheus::params::alerts,
+  Array $alert_relabel_config      = $::prometheus::params::alert_relabel_config,
+  Array $alertmanagers_config      = $::prometheus::params::alertmanagers_config,
+  String $storage_retention        = $::prometheus::params::storage_retention,
+  Array[Hash] $collect_scrape_jobs = $::prometheus::params::collect_scrape_jobs,
```
Comment (bastelfreak) [inline manifests/init.pp:168]: Datatypes \o/

File: manifests/scrape_job.pp [hunk 4]
```diff
@@ -0,0 +1,25 @@
+# TODO
+define prometheus::scrape_job (
+  String $job_name,
+  Array[String] $targets,
+  Hash[String, String] $labels = {},
+
```
Comment (bastelfreak) [inline manifests/scrape_job.pp]: please remove the newline here

File: manifests/scrape_job.pp [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+    }
+  ]
+  file { "${collect_dir}/${job_name}_${name}.yaml":
+    ensure  => present,
+    owner   => $user,
+    group   => $group,
+    mode    => $mode,
+    content => template("${module_name}/file_sd_config.yaml.erb"),
+  }
+}
```
Comment (bastelfreak) [inline manifests/scrape_job.pp]: please add anewline

**Comment by bastelfreak** [comment]
This looks good on a first view, but we need tests for it.

**Comment by viq** [comment]
What's the state of this? It would be really handy for me ;)

**Comment by costela** [comment]
@viq sorry, this kinda fell between the cracks.
I'll try to get to it next week, but I wouldn't complain if someone could conjure up some tests before that :wink:

**Comment by ajcollett** [comment]
I am super interested in this. Can you give me an idea of what still needs to happen? Not sure I can help with much, other than real world testing...

**Comment by costela** [comment]
@ajcollett sorry, I just haven't had the time to add the needed tests. And now I also have to rebase and solve conflicts... :unamused:
I *might* be able to squeeze this in a slot on Friday, but I can't make any promises.

**Comment by ajcollett** [comment]
No problem. I hope it's easier than it seems!

I am slightly surprised this particular issue isn't more popular. Is it just that people are all doing this themselves anyway?

**Comment by bastelfreak** [comment]
@ajcollett personally, I use consul for this.It already knows all services that I have, they expose a prometheus endpoint. Also I've got certain exporters that are also registered to consul. That scales way better than exported resources.

File: manifests/config.pp (hunk 1 above)
Comment (bastelfreak) [reply manifests/config.pp:6]: Do you know the structure of the Hash? Could that be enforced as well within the datatype?

File: manifests/daemon.pp [hunk 6]
```diff
@@ -78,6 +78,9 @@
   $manage_service                 = true,
   Hash[String, Scalar] $env_vars  = {},
   Optional[String] $env_file_path = $::prometheus::params::env_file_path,
+  Boolean $export_scrape_job      = false,
+  String $scrape_host             = $::fqdn,
```
Comment (bastelfreak) [inline manifests/daemon.pp:82]: Please use `$facts['fqdn']`

File: manifests/config.pp (hunk 1 above)
Comment (anarcat) [reply manifests/config.pp:6]: how about we just don't try to fix that in this specific pull request? there wasn't a type before either...
```suggestion
  $scrape_configs,
```

:P

File: manifests/init.pp [hunk 7]
```diff
@@ -113,50 +113,59 @@
 #  prometheus 1.8.*, only durations understood by golang's time.ParseDuration are supported. Starting
 #  with prometheus 2, durations can also be given in days, weeks and years.
 #
+#  [*collect_scrape_jobs*]
+#  Array of scrape_configs. Format, e.g.:
+#  - job_name: some_exporter
+#    scheme: https
+#  The jobs defined here will be used to collect resources exported via prometheus::daemon,
```
Comment (anarcat) [inline manifests/init.pp:120]: i find this datastructure a little over-complicated and inconvenient - what if i want to collect all jobs, regardless of the type? the way it's setup now, i have to specify each one by hand....

i'm thinking of simplifying this and simply make that a mixed type of boolean / array of job names. if it's a list, it behaves basically the same way (minus the scheme override, which belongs in the scrape_job def). if it's a boolean and true, all jobs are collected. if it's false, nothing is collected of course.

i'm working on an update of this PR to make it work with latest master which includes those changes.

Comment (costela) [reply manifests/init.pp:120]: Thanks for picking up the work!
I'm not sure I understand your suggestion though. I'll follow up on your new PR #304.

**Comment by anarcat** [comment]
okay, i have done my homework and this is actually working in production in #304 so I think this PR should be closed and discussion be moved there.

**Comment by costela** [comment]
agreed

**Pull request closed without merging**